src/backend/distributed/metadata/metadata_sync.c

```diff
@@ -4040,26 +4329,298 @@ ColocationGroupCreateCommandList(void)
 							 "NULL, NULL)");
 		}
 
-		colocationTuple = systable_getnext_ordered(scanDescriptor, ForwardScanDirection);
+		appendStringInfo(colocationGroupCreateCommand,
+						 ") SELECT pg_catalog.citus_internal_add_colocation_metadata("
+						 "colocationid, shardcount, replicationfactor, "
+						 "distributioncolumntype, coalesce(c.oid, 0)) "
+						 "FROM colocation_group_data d LEFT JOIN pg_collation c "
+						 "ON (d.distributioncolumncollationname = c.collname "
+						 "AND d.distributioncolumncollationschema::regnamespace"
+						 " = c.collnamespace)");
+
+		List *commandList = list_make1(colocationGroupCreateCommand->data);
+		SendOrCollectCommandListToActivatedNodes(context, commandList);
 	}
+	MemoryContextSwitchTo(oldContext);
+
+	systable_endscan(scanDesc);
+	table_close(relation, AccessShareLock);
+}
+
+
+/*
+ * SendDependencyCreationCommands sends dependency creation commands to workers
+ * with transactional or nontransactional mode according to transactionMode
+ * inside metadataSyncContext.
+ */
+void
+SendDependencyCreationCommands(MetadataSyncContext *context)
+{
+	/* disable ddl propagation */
+	SendOrCollectCommandListToActivatedNodes(context,
+											 list_make1(DISABLE_DDL_PROPAGATION));
 
-	systable_endscan_ordered(scanDescriptor);
-	index_close(colocationIdIndexRel, AccessShareLock);
-	table_close(pgDistColocation, AccessShareLock);
+	MemoryContext oldContext = MemoryContextSwitchTo(context->context);
 
-	if (!hasColocations)
+	/* collect all dependencies in creation order and get their ddl commands */
+	List *dependencies = GetDistributedObjectAddressList();
+
+	/*
+	 * Depending on changes in the environment, such as the enable_metadata_sync guc
+	 * there might be objects in the distributed object address list that should currently
+	 * not be propagated by citus as they are 'not supported'.
+	 */
+	dependencies = FilterObjectAddressListByPredicate(dependencies,
+													  &SupportedDependencyByCitus);
+
+	dependencies = OrderObjectAddressListInDependencyOrder(dependencies);
+
+	MemoryContext commandsContext = AllocSetContextCreate(context->context,
+														  "dependency commands context",
+														  ALLOCSET_DEFAULT_SIZES);
+	MemoryContextSwitchTo(commandsContext);
+	ObjectAddress *dependency = NULL;
+	foreach_ptr(dependency, dependencies)
 	{
-		return NIL;
+		if (IsAnyObjectAddressOwnedByExtension(list_make1(dependency), NULL))
+		{
+			/*
+			 * We expect extension-owned objects to be created as a result
+			 * of the extension being created.
+			 */
+			continue;
+		}
+
+		if (!MetadataSyncCollectsCommands(context))
```

Comment:
lets use ResetMetadataSyncMemoryContext

Reply:
Also, I'd probably do this at the top of the loop, not after IsAnyObjectAddressOwnedByExtension

Reply:
Added comment like:

We need to create a subcontext as we reset the context after each dependency creation but we want to preserve all dependency objects at metadataSyncContext.